<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Обзор-данных" data-toc-modified-id="Обзор-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Обзор данных</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Обработка-пропусков" data-toc-modified-id="Обработка-пропусков-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Обработка пропусков</a></span></li><li><span><a href="#Замена-типа-данных" data-toc-modified-id="Замена-типа-данных-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Замена типа данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Лемматизация" data-toc-modified-id="Лемматизация-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Лемматизация</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Вопросы-заказчика" data-toc-modified-id="Вопросы-заказчика-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Вопросы заказчика</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Есть ли зависимость между наличием детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

***Исследование надёжности заёмщиков***

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.  

*Описание данных*    
children — количество детей в семье  
days_employed — общий трудовой стаж в днях  
dob_years — возраст клиента в годах  
education — уровень образования клиента  
education_id — идентификатор уровня образования  
family_status — семейное положение  
family_status_id — идентификатор семейного положения  
gender — пол клиента  
income_type — тип занятости  
debt — имел ли задолженность по возврату кредитов  
total_income — ежемесячный доход  
purpose — цель получения кредита  

### Обзор данных

In [1]:
import pandas as pd
from pymystem3 import Mystem

In [2]:
try:
    df = pd.read_csv(r'C:\Users\Susanna\Downloads\data.csv')
except:
    df=pd.read_csv('/datasets/data.csv')

Посмотрим сколько есть пропусков:

In [3]:
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [4]:
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

<b>Вывод:</b>  
Пропуски есть в столбцах days_employed и total_income. Всего пропусков 2174 клиентов из 21525, это около 10 % всех клиентов. 
Предположим, что пропуски случайны.

### Предобработка данных

#### Обработка пропусков

В столбцах children, days_employed и total_income есть отрицательные знаения. Предполагаю что это резульат неправильной выгрузки данных
Уберем все минусы перед числами:

In [5]:
df['children']=df['children'].abs()
df['days_employed']=df['days_employed'].abs()
df['total_income']=df['total_income'].abs()

Заполним пропуски средним значением из каждой группы.

Найдем средние значения total_income для каждой группы income_type

In [6]:
log = df.groupby('income_type').agg({'days_employed':['mean'], 'total_income': ['mean']}).reset_index()
log.set_axis(['income_type', 'days_employed_mean', 'total_income_mean'], axis='columns', inplace=True)
log=log.sort_values(by='income_type')

print(log)

       income_type  days_employed_mean  total_income_mean
0      безработный       366413.652744      131339.751676
1        в декрете         3296.759962       53829.130729
2      госслужащий         3399.896902      170898.309923
3        компаньон         2111.524398      202417.461462
4        пенсионер       365003.491245      137127.465690
5  предприниматель          520.848083      499163.144947
6        сотрудник         2326.499216      161380.260488
7          студент          578.751554       98201.625314


Создадим ф-ии для замены пропусков:

In [7]:
def func(row):
    if row['days_employed']==row['days_employed']:
        return row['days_employed']
    else:
        if row['income_type']=='сотрудник':
            return log.loc[6, 'days_employed_mean']
        elif row['income_type']=='компаньон':
            return log.loc[3, 'days_employed_mean']
        elif row['income_type']=='пенсионер':
            return log.loc[4, 'days_employed_mean']
        elif row['income_type']=='госслужащий':
            return log.loc[2, 'days_employed_mean']
        elif row['income_type']=='предприниматель':
            return log.loc[5, 'days_employed_mean']

In [8]:
def func2(row):
    if row['total_income']==row['total_income']:
        return row['total_income']
    else:
        if row['income_type']=='сотрудник':
            return log.loc[6, 'total_income_mean']
        elif row['income_type']=='компаньон':
            return log.loc[3, 'total_income_mean']
        elif row['income_type']=='пенсионер':
            return log.loc[4, 'total_income_mean']
        elif row['income_type']=='госслужащий':
            return log.loc[2, 'total_income_mean']
        elif row['income_type']=='предприниматель':
            return log.loc[5, 'total_income_mean']

In [9]:
df['days_employed']=df.apply(func, axis=1)
df['total_income']=df.apply(func2, axis=1)

Проверим как удалились пропуски

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<b>Вывод:</b>   
Пропусков больше нет.

#### Замена типа данных

Заменим вещественный тип данных на целочисленный:

In [11]:
df['days_employed']=df['days_employed'].astype('int')
df['total_income']=df['total_income'].astype('int')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  int64 
 2   dob_years         21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  int64 
 11  purpose           21525 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


<b> Вывод:</b>   
Для замены вещественного типа данных на целочисленный был использован метод .astype

#### Обработка дубликатов

Для того чтобы в дальнейем корректно удалить все дубликаты
в столбцах days_employed, family_status, gender, income_type и purpose переведем все буквы в строчные:

In [12]:
df['education']=df['education'].str.lower()
df['family_status']=df['family_status'].str.lower()
df['gender']=df['gender'].str.lower()
df['income_type']=df['income_type'].str.lower()
df['purpose']=df['purpose'].str.lower()

Просмотрм все уникальные значения столбцов, чтобы убедиться в отсутсвии ошибок:

In [13]:
print(df['children'].unique())
print(df['education'].unique())
print(df['education_id'].unique())
print(df['family_status'].unique())
print(df['family_status_id'].unique())
print(df['gender'].unique())
print(df['income_type'].unique())
print(df['debt'].unique())
print(df['purpose'].unique())

[ 1  0  3  2  4 20  5]
['высшее' 'среднее' 'неоконченное высшее' 'начальное' 'ученая степень']
[0 1 2 3 4]
['женат / замужем' 'гражданский брак' 'вдовец / вдова' 'в разводе'
 'не женат / не замужем']
[0 1 2 3 4]
['f' 'm' 'xna']
['сотрудник' 'пенсионер' 'компаньон' 'госслужащий' 'безработный'
 'предприниматель' 'студент' 'в декрете']
[0 1]
['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'операции со своей недвижимостью' 'автомобили' 'заняться образованием'
 'сделка с подержанным автомобилем' 'получение образования' 'автомобиль'
 'свадьба

In [14]:
df['gender'].value_counts()

f      14236
m       7288
xna        1
Name: gender, dtype: int64

появилось странное значение xna, оно всего одно, поэтому я удалю эту строку

In [15]:
df = df.loc[df['gender'] != 'xna']

Удалим все дубликаты

In [16]:
df=df.drop_duplicates().reset_index(drop= True)

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21453 entries, 0 to 21452
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21453 non-null  int64 
 1   days_employed     21453 non-null  int64 
 2   dob_years         21453 non-null  int64 
 3   education         21453 non-null  object
 4   education_id      21453 non-null  int64 
 5   family_status     21453 non-null  object
 6   family_status_id  21453 non-null  int64 
 7   gender            21453 non-null  object
 8   income_type       21453 non-null  object
 9   debt              21453 non-null  int64 
 10  total_income      21453 non-null  int64 
 11  purpose           21453 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


<b> Вывод:</b>   
После обработки дубликотов осталось 21453 строки. Для удаления дубликатов использовался метод drop_duplicates()

#### Лемматизация

Выделим леммы в значениях столбца с целями получения кредита

In [18]:
m = Mystem()

def lemmatization(row):
    lemmas= m.lemmatize(row['purpose'])
    return lemmas

Добавим столбец с леммами

In [19]:
df['purpose_lemmatized']=df.apply(lemmatization, axis=1)
print(df.head())

   children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      f   сотрудник     0        253875   
1   женат / замужем                 0      f   сотрудник     0        112080   
2   женат / замужем                 0      m   сотрудник     0        145885   
3   женат / замужем                 0      m   сотрудник     0        267628   
4  гражданский брак                 1      f   пенсионер     0        158616   

                      purpose                    purpose_lemmatized  
0               покупка жилья               [покупка,  , ж

<b> Вывод:</b>  
Выделены леммы в столбце с целями. Видно что много одинаковых целей, записанных по-разному. Необходимо категоризировать столбец с целями.

#### Категоризация данных

Категоризируем столбец с целями получения кредита

In [20]:
def categoryzation(row):
    i=row['purpose_lemmatized']
    if 'жилье' in i:
        return 'недвижимость'
    if 'недвижимость' in i:
        return 'недвижимость'
    if 'автомобиль' in i:
        return 'автомобиль'
    if 'образование' in i:
        return 'образование'
    if 'свадьба' in i:
        return 'свадьба'
    else:
        return 'другое'

Добавим столбец с категриями целей получения кредита

In [21]:
df['purpose_category']=df.apply(categoryzation, axis=1) 
df.head()

children  days_employed  dob_years education  education_id  \
0         1           8437         42    высшее             0   
1         1           4024         36   среднее             1   
2         0           5623         33   среднее             1   
3         3           4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      f   сотрудник     0        253875   
1   женат / замужем                 0      f   сотрудник     0        112080   
2   женат / замужем                 0      m   сотрудник     0        145885   
3   женат / замужем                 0      m   сотрудник     0        267628   
4  гражданский брак                 1      f   пенсионер     0        158616   

                      purpose                    purpose_lemmatized  \
0               покупка жилья               [покупка,  , жилье, \n]   
1     приобретение автомобиля     [приобретение,  , автомобиль, \n]   
2               покупка жилья               [покупка,  , жилье, \n]   
3  дополнительное образование  [дополнительный,  , образование, \n]   
4             сыграть свадьбу             [сыграть,  , свадьба, \n]   

  purpose_category  
0     недвижимость  
1       автомобиль  
2     недвижимость  
3      образование  
4          свадьба

<b> Вывод:</b>  
Объединили похожие цели в категории.

###  Вопросы заказчика

#### Есть ли зависимость между наличием детей и возвратом кредита в срок?

Посмотрим сколько каких значений есть в столбце debt:

In [22]:
df['debt'].value_counts()

0    19712
1     1741
Name: debt, dtype: int64

Посчитаем отношение общего числа детей у тех кто имел задолженности по кредиту к общему числу самих должников.   
И такое же отношение у тех кто платил по кредиту вовремя.

In [23]:
a=df[df['debt']==1]['children'].sum()/df[df['debt']==1]['debt'].count()
b=df[df['debt']==0]['children'].sum()/df[df['debt']==0]['debt'].count()
print('В среднем на человека имевшего задолженности по возврату кредита приходится детей: {:.2f}'.format(a))
print('В среднем на человека НЕ имевшего задолженности по возврату кредита приходится детей: {:.2f}'.format(b))

В среднем на человека имевшего задолженности по возврату кредита приходится детей: 0.63
В среднем на человека НЕ имевшего задолженности по возврату кредита приходится детей: 0.54


<b> Вывод: </b>   
Тех, кто имел задолженности по возвату кредита сильно меньше. Количество детей на человека почти одинаково, однако у тех, кто никогда не имел задолженностей по кредитам их немного меньше.
Есть ли зависимость между наличием детей и возвратом кредита в срок? - Есть. Чем меньше детей, тем лучше люди платят по кредитам.

#### Есть ли зависимость между семейным положением и возвратом кредита в срок?

Посмотрим количесво людей в разных семейных положениях:

In [24]:
df['family_status'].value_counts()

женат / замужем          12339
гражданский брак          4150
не женат / не замужем     2810
в разводе                 1195
вдовец / вдова             959
Name: family_status, dtype: int64

Разделим всех клиентов на тех, кто имел задолженности, и тех, кто их не имел, в зависимость от семейного положения:

In [25]:
df_pivot = df.pivot_table(index=['family_status_id', 'family_status'], columns='debt', values='children', aggfunc='count').reset_index()

df_pivot.set_axis(['family_status_id', 'family_status', 'ne_imeli_zadol', 'imeli_zadol'], axis='columns', inplace=True)
df_pivot['ratio']=df_pivot['ne_imeli_zadol']/df_pivot['imeli_zadol']
df_pivot.sort_values(by = 'ratio',ascending = False)

family_status_id          family_status  ne_imeli_zadol  imeli_zadol  \
2                 2         вдовец / вдова             896           63   
3                 3              в разводе            1110           85   
0                 0        женат / замужем           11408          931   
1                 1       гражданский брак            3762          388   
4                 4  не женат / не замужем            2536          274   

       ratio  
2  14.222222  
3  13.058824  
0  12.253491  
1   9.695876  
4   9.255474

<b> Вывод:</b>  
Есть ли зависимость между семейным положением и возвратом кредита в срок? - Есть. Лучше всего расплачивались по кредитам вдовы/вдовцы, а хуже всего не женатые/не замужние. Чем выше ratio, тем лучше эти категории отдавали кредиты вовремя. 

#### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Посмотрим максимальную, минимальную и медианную зарплаты клиентов

In [26]:
print('Максимальная зарплата среди клиентов: {}'.format(df['total_income'].max()))
print('Минимальная зарплата среди клиентов: {}'.format(df['total_income'].min()))
median_income=df['total_income'].median()
print('Средняя зарплата среди клиентов: {}'.format(median_income))

Максимальная зарплата среди клиентов: 2265604
Минимальная зарплата среди клиентов: 20667
Средняя зарплата среди клиентов: 151876.0


Разделим условно зарплаты всех клиентов на "высокую", "среднюю" ,"низкую" и "нет зарплаты".

In [27]:
def income_group(row):
        i=row['total_income']
        if i == 0:
                return 'нет зарплаты'
        if i < median_income:
                return 'низкая зрплата'
        return 'высокая зарплата'

df['income_group'] = df.apply(income_group, axis=1)

In [28]:
df_pivot2 = df.pivot_table(index=['income_group'], columns='debt', values='children', aggfunc='count')
df_pivot2 = df_pivot2.reset_index()
df_pivot2.set_axis(['income_group', 'ne_imeli_zadol', 'imeli_zadol'], axis='columns', inplace=True)
df_pivot2['ratio']=df_pivot2['imeli_zadol']/df_pivot2['ne_imeli_zadol']
df_pivot2.sort_values(by = 'ratio',ascending = False)

income_group  ne_imeli_zadol  imeli_zadol     ratio
1    низкая зрплата            9824          902  0.091816
0  высокая зарплата            9888          839  0.084850

<b> Вывод:</b>  
Есть ли зависимость между уровнем дохода и возвратом кредита в срок? - Нет. Клиенты с любой категорией зарплаты выплачивали кредиты примерно одинаоково.

#### Как разные цели кредита влияют на его возврат в срок?

In [29]:
df['purpose_category'].value_counts()

недвижимость    10810
автомобиль       4306
образование      4013
свадьба          2324
Name: purpose_category, dtype: int64

In [30]:
df_pivot3 = df.pivot_table(index=['purpose_category'], columns='debt', values='children', aggfunc='count').reset_index()

df_pivot3.set_axis(['purpose_category', 'ne_imeli_zadol', 'imeli_zadol'], axis='columns', inplace=True)
df_pivot3['ratio']=df_pivot3['imeli_zadol']/df_pivot3['ne_imeli_zadol']
df_pivot3.sort_values(by = 'ratio',ascending = False)

purpose_category  ne_imeli_zadol  imeli_zadol     ratio
0       автомобиль            3903          403  0.103254
2      образование            3643          370  0.101565
3          свадьба            2138          186  0.086997
1     недвижимость           10028          782  0.077982

<b> Вывод:</b>  
Клиенты, которые брали кредит для преобретения или ремонта недвижимости, реже имеют проблемы с возвратом, чем те, кто брал кредит на автомобиль или образование.

### Общий вывод
Тех, кто имел задолженности по возвату кредита сильно меньше. 
Большинство клиентов платит по кредитам вовремя.
Количество детей слабо влияет на способность вовремя возвращать деньги. Однако с появлением детей клиенты немного хуже справляются с выплатами в срок.
Лучше всего расплачивались по кредитам вдовы/вдовцы, а хуже всего не женатые/не замужние.
Размер зарплаты не влияет на способность вовремя оплачвать кредит.
Клиенты, которые брали кредит для преобретения или ремонта недвижимости, реже имеют проблемы с возвратом, чем те, кто брал кредит на автомобиль или образование.